enhancements/insights/prometheus-alerts-for-insights-recommendations.md

```diff
@@ -0,0 +1,268 @@
+- Configuration option allowing the user to disable/enable this functionality.
+  Since this feature consumes resources on both sides (cluster and Smart Proxy),
+  it should only be enabled on clusters where the user actually wants to use it. (It should be disabled by default, unless the user explicitly enables it.)
+- The enabling should be done using the same approach as the rest of the
+  Insights Operator configuration (i.e., currently using the `support` secret).
+
+### Non-Goals
+
+- Insights Operator fetching additional data, from the Smart Proxy or any other source.
+- Modifications or additions to the user notification systems (the goal is to
+  utilize the existing notification systems based on Prometheus alerts - most
+  notably Alertmanager).
+
+## Proposal
+
+### User Stories
+
+As a user, I want to be immediately notified about a recommendation that has
+become active on my cluster without having to periodically check the Insights
+tab. I want to receive notifications through my preferred channel (e.g., Slack,
+email) configured via Alertmanager when a recommendation becomes active, and
+another one when it becomes inactive again.
+
+### Why should this be implemented in Insights Operator?
+
+The Insights Operator is already fetching the relevant data from the Smart Proxy,
+and managing Prometheus metrics based on this data. Therefore, it makes sense
+for additional information about recommendations to be implemented alongside
+what already exists.
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints [optional]
+
+Please note that IO refers to the Insights Operator in this context.
+
+1. IO is enabled and periodically uploads archives to the Ingress service,
+   which are then processed by the CCX data pipeline. (standard behavior)
+2. After a successful archive upload, IO waits for the report to become
+   available, fetches it from the Smart Proxy and processes it. (This is an
+   extension of the current behavior, where we only look at the number of
+   active recommendations per severity level.) (needs to be implemented)
+3. In-cluster Prometheus metrics indicating the status of recommendations
+   being active are set accordingly.
+    - If a recommendation was previously registered as active and is present
+      in the latest Insights report, IO performs no additional action for this
+      recommendation.
+    - If a recommendation was previously registered as active and the latest
+      report no longer contains it, the metric will either be reset or
+      unregistered from the Prometheus instance if possible. (remove alerts for
+      no longer hitting recommendations)
+    - If a recommendation was previously registered as inactive or the metric
+      for this recommendation is absent and the latest Insights report contains
+      it, a metric will be updated/created to indicate that the recommendation
+      is currently hitting. (indicate new recommendations)
```

Comment:
I think all of this can be collapsed into "Insights operator serves a metric time-series for any recommendations in the most recently retrieved report", and the alert rule you layer on top can handle do some aggregation to bridge outages due to the insights operator pod rolling, etc.  For example, [here][1] is the cluster-version operator looping over ClusterOperator conditions and writing a time-series with different labels for each hit, and [here][2] is the `ClusterOperatorDegraded` rule built on top of that metric.  openshift/cluster-version-operator#643 was an attempt to make a similar alert robust to CVO-pod restarts, but I think we need to add some time aggregation to that still to actually get it working.

[1]: https://github.com/openshift/cluster-version-operator/blob/0e9bc4ef03004fcc2bb0c58d39e5b49445a9f8f6/pkg/cvo/metrics.go#L478-L489
[2]: https://github.com/openshift/cluster-version-operator/blob/0e9bc4ef03004fcc2bb0c58d39e5b49445a9f8f6/install/0000_90_cluster-version-operator_02_servicemonitor.yaml#L88-L100

Reply:
Some context on [the metrics architecture][1], in case that's helpful (I'm not monitoring, so I might be wrong):

1. You declare a ServiceMonitor, pointing at your Service
2. Prometheus notices the ServiceMonitor, and periodically stops by to scrape any metrics you are serving.
3. If the scrape fails, Prometheus will extrapolate/interpolate to paper over small gaps.
3. If the scrape succeeds, any metrics you are currently serving get stored in the Prom database, and any metrics you are no longer serving get removed from the Prom database.

So:

1. You get a report.
2. You set up `insights_recommendation{someLabel="A",...} 1`, `insights_recommendation{someLabel="B",...} 1`, etc., based on the things you saw in that report.
3. Prom comes by and scrapes those up.
4. Prom comes by again, for another round of scraping.
5. Fetch a new report!  You set up `insights_recommendation{someLabel="A",...} 1`, `insights_recommendation{someLabel="C",...} 1`.  No mention of `B` in this report.  You don't have to clear anything.
6. Prom comes by and scrapes those up.  Hey, a new `C`!  Hmm, I guess `B` is gone.

As long as you aren't needlessly churning label values on the metrics you're serving, you won't be triggering notification churn.  And if I/O crashes halfway through a metrics update, you'll be starting over with a fresh report-fetch and metrics setup, and as far as Prom is concerned that just looks like a healthy pod/container roll.

[1]: https://prometheus.io/docs/introduction/overview/#components

Reply:
I don't really understand what you mean by not having to clear anything. The

Reply:
I'm a bit fuzzy on how Prom's Go metrics server is set up, but the way it seems to work in the CVO is:

1. Something calls `Collect` on us periodically (a metrics-serving goroutine?) to ask what metrics it should currently be serving.
2. Inside `Collect`, we [push into the provided channel][1] anything we thing we should be serving right now.
3. The metrics server or whatever called `Collect` updates its internal cache to match.

Periodically Prom's scraper comes around, and gets whatever was in the metrics-server's cache at that time.  So no CVO-specific steps to clear anything on the metrics-serving side either.  That's magically happening in vendored code in step 3.

[1]: https://github.com/openshift/cluster-version-operator/blob/0e9bc4ef03004fcc2bb0c58d39e5b49445a9f8f6/pkg/cvo/metrics.go#L478-L489

Reply:
That's really interesting. We don't implement any `Collect` like this AFAIK (we only do [this](https://github.com/openshift/insights-operator/blob/master/pkg/insights/insightsclient/insightsclient.go#L245)) and I am wondering if we should. The interface is defined here https://github.com/prometheus/client_golang/blob/main/prometheus/collector.go#L27

Reply:
Okay, I see what you mean, more or less, in the CVO code. Unfortunately, that's not how IO manages it metrics at the moment, so we will probably have to either refactor the IO code to follow this pattern or work around it in some other way, as was originally suggested in the enhancement. From my understanding, and I'll have to double-check this, IO simply sets the metrics ad-hoc, and doesn't really have this `Collect` callback at all.

Reply:
I would expect that the insights operator implements the custom collector pattern with const metrics. The operator will probably have an internal representation of the recommendations and it needs to translate them into metrics whenever Prometheus scraping happens. See https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics for more details.
Trying to use the `GaugeVec`, `CounterVec`, ... types is prone to errors because their interfaces don't make it easy to remove labels.

Reply:
> Trying to use the `GaugeVec`, `CounterVec`, ... types is prone to errors because their interfaces don't make it easy to remove labels.

Will we need to remove labels for this use-case?  It sounded like we'd have a recommendation ID, name/slug, active-ness, and detail URI.  And that we'd have those same labels in every exported entry.  That sounds pretty much identical to the CVO's `cluster_operator_conditions` to me, and we are using `NewGaugeVec` for that, and exporting information for multiple conditions for multiple operators simultaneously.

Reply:
So the suggestion is to use individual const metrics for each recommendation, with their respective labels? I don't know if removing labels, as you mention, is something we need to do in IO.

I'm not really very familiar with const metrics because we have not used this Collector pattern before in IO. Would it be okay to use a gauge type with the const metrics and store the timestamp of the most recent report with the recommendation in it? It's not exactly clear to me from the documentation if it's only "const" in the context of how it's created or if the Prometheus service will prevent us from changing it once it has been created (because with each report, we may need to update that timestamp).

Then again, I don't think it makes sense to store the timestamp in each metric because their values would all be the same. We could stick to the binary metrics for recommendations and have another metric contain the timestamp of the last report. Since every recommendation was either in the latest report or it isn't active, this should not be an issue.